File: python/chimes.py
import math

import numpy as np
import matplotlib.pyplot as plt
import scipy.signal as sig
from scipy.io.wavfile import write as wav_write

# An attempt at recreating the results from the paper:
# "Modal Synthesis of Wind Chime Sounds with Stochastic
#  Event Triggering" by Teemu Lukkari and Vesa Välimäki

FS = 11025   # [Hz]
TS = 1 / FS  # [s]
END_TIME = 25.0  # [s]
N_BELLS = 5

T_60 = [40, 7, 2, 1, 0.5]
Rs = list(map(lambda t: math.pow(1e-6, 1 / (t * FS)), T_60))
Gs = [0.0787, 0.1849, 1.0000, 0.0136, 0.0275]

# modes for 5 tubes
MODES = [[219.8, 590.2, 1115.4, 1766.2, 2513.9],
         [245.8, 657.6, 1239.2, 1955.3, 2773.1],
         [293.9, 782.4, 1465.2, 2311.8, 3278.8],
         [331.6, 875.5, 1633.0, 2576.5, 3654.2],
         [366.1, 967.5, 1794.2, 2831.0, 4015.1]]

assert(len(MODES) == N_BELLS)
assert(len(Rs) == len(MODES[0]))
assert(len(Gs) == len(MODES[0]))
assert(len(T_60) == len(MODES[0]))


class Trigger:
    def __init__(self, n_bells):
        self.n_bells = n_bells
        self.s_low = round(0.03 / TS)
        self.s_high = round(0.05 / TS)
        self.next_s = 0
        self.state = 0

    def update(self, n):
        p = hit_prob(energy(n))
        if n == self.next_s:
            if self.state == 0:
                ps = [1 - p] + ([p / self.n_bells] * self.n_bells)
                self.state = np.random.choice(np.arange(0, self.n_bells + 1),
                                              p=ps)
            else:
                cs = [0, self.state - 1 if self.state > 1 else self.n_bells,
                      self.state + 1 if self.state < self.n_bells else 1]
                ps = [1 - p, p / 2, p / 2]
                self.state = np.random.choice(cs, p=ps)
            self.next_s = n + \
                np.random.choice(range(self.s_low, self.s_high + 1, 1))
        return self.state


class BurstGen:
    def __init__(self):
        self.R_input = 0.97
        self.E = 0.0
        self.start = 0
        self.n = 0
        self.active = False

    def update(self, n):
        self.n = n
        a = math.sqrt(self.E + 0.1)
        if self.active:
            if (n - self.start) <= (0.02 / TS):
                d = a * math.pow(self.R_input, n - self.start)
                s = np.random.uniform(-1, 1, 1)[0]
                return s * d
            else:
                self.active = False
                return 0.0
        else:
            return 0.0

    def reset(self, E):
        self.E = E
        self.start = self.n
        self.active = True


def wind(t):
    if t < 2:
        return 0.0
    elif t < 10:
        return 2 * (t - 2) / 8
    elif t < 20:
        return 2.0
    else:
        return 0.0


def energy(n):
    Rdec = 0.9999
    dn = TS * wind((n - 1) * TS)
    En = Rdec * (energy.En_prev + dn)
    energy.En_prev = En
    return En


def hit_prob(e):
    return 1.0 / (1 + (99 * math.exp(-2 * e)))


def resonator_coefs(modes):
    fs = []
    for i, f in enumerate(modes):
        R = Rs[i]
        G = Gs[i]
        cos_theta = math.cos(2 * math.pi * f * TS) * (2 * R) / (1 + (R * R))
        sin_theta = math.sqrt(1 - (cos_theta * cos_theta))
        A_zero = (1 + (R * R)) * sin_theta
        b_s = [A_zero * G, 0.0, -A_zero * G]
        a_s = [1.0, -2 * R * cos_theta, R * R]
        fs.append((b_s, a_s))
    return fs


def write_lines_to_file(ls, fn):
    with open(fn, "w", encoding='utf-8') as f:
        for l in ls:
            f.write(l + "\n")


def gen_C_cfg():
    n_reson = len(MODES[0])

    ls = ["#ifndef __CHIMES_CFG__",
          "#define __CHIMES_CFG__",
          "",
          "#define N_TUBES ({})".format(N_BELLS),
          "#define N_RESON ({})".format(n_reson),
          "",
          "typedef const float (*a_coefs_t)[3];",
          "typedef const float (*b_coefs_t)[3];",
          ""]

    for n in range(N_BELLS):
        f_coefs = resonator_coefs(MODES[n])

        for i, (b_s, a_s) in enumerate(f_coefs):
            ls.append("static const float T{}_B{}[{}] = {{ {} }};".format(
                n + 1, i + 1, len(b_s), ", ".join(list(map(str, b_s)))))
            ls.append("static const float T{}_A{}[{}] = {{ {} }};".format(
                n + 1, i + 1, len(a_s), ", ".join(list(map(str, a_s)))))
            ls.append("")

    def gen_init(ab):
        return ", ".join(["&T##_x##_{}{}".format(ab, i + 1) for i in range(n_reson)])

    ls.extend(
        ["#define TUBE_INIT(_x) \\",
         "    do \\",
         "    { \\",
         "        tube_init(&chime ->tubes[_x - 1], \\",
         "                  &(b_coefs_t[N_RESON]){" + gen_init("B") + "},  \\",
         "                  &(a_coefs_t[N_RESON]){" + gen_init("A") + "}); \\",
         "    } while (0)",
         ""])

    ls.extend(
        ["#define TUBES_INIT() \\",
         "    do \\",
         "    { \\"])

    ls.extend(["        TUBE_INIT({}); \\".format(i + 1)
              for i in range(N_BELLS)])

    ls.extend(
        ["    } while (0)",
         ""])

    ls.extend([
        "#define TRIG_PROBS() \\",
        "    const float ps[N_TUBES + 1] = {1.0 - p, " +
        ", ".join(["p / N_TUBES" for _ in range(N_BELLS)]) + "}; \\",
        "    const uint8_t es[N_TUBES + 1] = {0, " +
        ", ".join([str(i + 1) for i in range(N_BELLS)]) + "}; \\",
        ""])

    ls.append("#endif // __CHIMES_CFG__")
    write_lines_to_file(ls, "../include/chimes_cfg.h")


time = np.arange(0, END_TIME, TS)
wnd_s = np.array(list(map(wind, time)))
energy.En_prev = 0
en = list(map(energy, range(len(time))))
h_prob = list(map(hit_prob, en))

trigger = Trigger(N_BELLS)
energy.En_prev = 0
strikes = [trigger.update(i) for i in range(len(time))]
strikes = list(
    map(lambda s: 0 if s[0] == s[1] else s[0],
        zip(strikes, strikes[1:] + [0])))

for i, ms in enumerate(MODES):
    n = FS
    bg = BurstGen()
    bg.reset(0.2)
    brst = [bg.update(j) for j in range(n)]
    f_coefs = resonator_coefs(ms)

    output = np.sum(
        np.array([sig.lfilter(b_s, a_s, brst) for (b_s, a_s) in f_coefs]), axis=0) / len(f_coefs)

    # save the synthesized tube samples
    wav_write("wav/tube_{}.wav".format(i + 1), FS, output)

chimes = np.zeros(len(strikes))

for n in range(N_BELLS):
    bell_n = list(map(lambda s: s == n + 1, strikes))
    bg = BurstGen()
    brsts = []

    for i, s in enumerate(bell_n):
        if s:
            bg.reset(en[i])
        brsts.append(bg.update(i))

    output = np.zeros(len(bell_n))
    f_coefs = resonator_coefs(MODES[n])

    for i, (b_s, a_s) in enumerate(f_coefs):
        output += sig.lfilter(b_s, a_s, brsts)

    chimes += output / (N_BELLS * len(f_coefs))

gen_C_cfg()

fig, axs = plt.subplots(5)
fig.set_figwidth(20)
fig.set_figheight(10)

for ax in axs:
    ax.grid(True)
    ax.set_xlim(0, END_TIME)
    ax.set_xticks(np.arange(min(time), max(time)+1, 1.0))

color = '#e08764'
axs[0].plot(time, wnd_s, color=color)
axs[0].set_ylabel("Wind force")

axs[1].plot(time, en, color=color)
axs[1].set_ylabel("Energy")

axs[2].plot(time, h_prob, color=color)
axs[2].set_ylabel("Hit probability")

loc_bars = list(filter(lambda p: p[1] > 0, zip(time, strikes)))
axs[3].bar([p[0] for p in loc_bars], [p[1] for p in loc_bars],
           width=0.08, color=color)
axs[3].set_ylabel("Strikes")

axs[4].plot(time, chimes, color=color)
axs[4].set_ylabel("Model output")

plt.xlabel("time [s]")
plt.show()

wav_write("wav/chimes.wav", FS, chimes * 0.9)
fig.savefig("plots.svg", format="svg", dpi=1200, transparent=True)

File: src/chimes.c
#include <stdint.h>
#include <stdio.h>
#include <stddef.h>
#include <stdlib.h>
#include <stdbool.h>

#include <math.h>

#include <assert.h>

#include "chimes.h"
#include "chimes_cfg.h"

#define TS (1.0 / SAMPLERATE)
#define WIND_POINTS (5)

typedef struct
{
    uint32_t timestamps[WIND_POINTS];
    uint32_t intervals[WIND_POINTS];
    float levels[WIND_POINTS];
} wind_t;

typedef struct
{
    wind_t wind;
    float E;
    uint32_t s_low;
    uint32_t s_high;
    uint32_t next_s;
    uint8_t state;
    uint8_t prev_state;
} trig_t;

typedef struct
{
    float R_input;
    float a;
    uint32_t start;
    uint32_t n;
    bool active;
} burst_t;

typedef struct
{
    b_coefs_t bs;
    a_coefs_t as;
    float xs[3];
    float ys[2];
} iirfilt_3_3_t;

typedef struct
{
    burst_t excitation;
    iirfilt_3_3_t resonators[N_RESON];
} tube_t;

struct _chime_t
{
    trig_t trig;
    tube_t tubes[N_TUBES];
    uint32_t i;
    uint32_t end;
};

static chime_t g_chime;

static float hit_prob(float e)
{
    return 1.0 / (1.0 + (99.0 * expf(-2.0 * e)));
}

static uint8_t rnd_choice(uint8_t const *const elems, float const *const probs, size_t len)
{
    assert(len > 0);
    float s = (float)rand() / (float)(RAND_MAX);
    float cp = 0.0;
    uint8_t ret = 0;

    for (size_t i = 0; i < len; i++)
    {
        cp += probs[i];

        if (s <= cp)
        {
            ret = elems[i];
            break;
        }
    }

    return ret;
}

static uint32_t wind_init(wind_t *wind,
                          float (*intervals)[WIND_POINTS],
                          float (*levels)[WIND_POINTS])
{
    uint32_t s = 0;

    for (size_t i = 0; i < WIND_POINTS; i++)
    {
        wind->intervals[i] = (*intervals)[i] * SAMPLERATE;
        s += wind->intervals[i];
        wind->timestamps[i] = s;
        wind->levels[i] = (*levels)[i];
    }

    return wind->timestamps[WIND_POINTS - 1];
}

static float wind_update(wind_t *wind, uint32_t i)
{
    for (size_t j = 0; j < WIND_POINTS; j++)
    {
        if (i < wind->timestamps[j])
        {
            uint32_t prev_tmstmp = j == 0 ? 0 : wind->timestamps[j - 1];
            float prev_level = j == 0 ? 0.0 : wind->levels[j - 1];

            float slope = (wind->levels[j] - prev_level) / (wind->intervals[j]);
            float level = prev_level + (slope * (i - prev_tmstmp));
            return level;
        }
    }

    return 0.0;
}

static float energy(wind_t *wind, uint32_t n)
{
    static float En_prev = 0.0;
    float Rdec = 0.9999;
    float dn = TS * wind_update(wind, n - 1);
    float En = Rdec * (En_prev + dn);
    En_prev = En;

    return En;
}

static float get_rand_up(float up)
{
    return ((float)rand() / (float)(RAND_MAX / up));
}

static void trig_init(trig_t *trig)
{
    trig->E = 0.0;
    trig->s_low = roundf(0.04 / TS);
    trig->s_high = roundf(0.06 / TS);

    trig->next_s = 0;
    trig->state = 0;
    trig->prev_state = 0;
}

static uint8_t trig_update(trig_t *trig, uint32_t i)
{
    trig->E = energy(&trig->wind, i);
    float p = hit_prob(trig->E);

    if (i == trig->next_s)
    {
        if (trig->state == 0)
        {
            TRIG_PROBS();
            trig->state = rnd_choice(es, ps, N_TUBES + 1);
        }
        else
        {
            const float ps[3] = {1.0 - p, p / 2.0, p / 2.0};
            const uint8_t es[3] = {
                0,
                trig->state > 1 ? trig->state - 1 : N_TUBES,
                trig->state < N_TUBES ? trig->state + 1 : 1,
            };
            trig->state = rnd_choice(es, ps, 3);
        }
        trig->next_s = i + (rand() % (trig->s_high - trig->s_low + 1)) + trig->s_low;
    }

    uint8_t ret = trig->state != trig->prev_state ? trig->state : 0;
    trig->prev_state = trig->state;

    return ret;
}

static float trig_get_energy(trig_t *trig)
{
    return trig->E;
}

static void burst_init(burst_t *burst)
{
    burst->R_input = 0.97;
    burst->a = 0.0;
    burst->start = 0;
    burst->n = 0;
    burst->active = false;
}

static void burst_start(burst_t *burst, float E)
{
    burst->a = sqrtf(E + 0.1);
    burst->start = burst->n;
    burst->active = true;
}

static float burst_update(burst_t *burst, uint32_t i)
{
    burst->n = i;

    if (burst->active)
    {
        if ((i - burst->start) <= (0.02 / TS))
        {
            float d = burst->a * powf(burst->R_input, i - burst->start);
            float s = get_rand_up(2.0) - 1.0;
            return s * d;
        }
        else
        {
            burst->active = false;
            return 0.0;
        }
    }
    else
    {
        return 0.0;
    }
}

static void iirfilt_3_3_init(iirfilt_3_3_t *iirfilt,
                             b_coefs_t bs,
                             a_coefs_t as)
{
    iirfilt->bs = bs;
    iirfilt->as = as;
    iirfilt->xs[0] = 0.0;
    iirfilt->xs[1] = 0.0;
    iirfilt->xs[2] = 0.0;
    iirfilt->ys[0] = 0.0;
    iirfilt->ys[1] = 0.0;
}

static float iirfilt_3_3_update(iirfilt_3_3_t *iirfilt, float x)
{
    iirfilt->xs[0] = iirfilt->xs[1];
    iirfilt->xs[1] = iirfilt->xs[2];
    iirfilt->xs[2] = x;

    float y = ((*iirfilt->bs)[0] * iirfilt->xs[2]) +
              // ((*iirfilt->bs)[1] * iirfilt->xs[1]) +
              ((*iirfilt->bs)[2] * iirfilt->xs[0]) -
              ((*iirfilt->as)[1] * iirfilt->ys[1]) -
              ((*iirfilt->as)[2] * iirfilt->ys[0]);

    // y /= (*iirfilt->as)[0];
    iirfilt->ys[0] = iirfilt->ys[1];
    iirfilt->ys[1] = y;

    return y;
}

static void tube_init(tube_t *tube,
                      b_coefs_t (*bs)[N_RESON],
                      a_coefs_t (*as)[N_RESON])
{
    for (size_t i = 0; i < N_RESON; i++)
    {
        iirfilt_3_3_init(&tube->resonators[i], (*bs)[i], (*as)[i]);
    }
    burst_init(&tube->excitation);
}

static float tube_update(tube_t *tube, uint32_t i, bool trig, float E)
{
    if (trig)
    {
        burst_start(&tube->excitation, E);
    }

    float x = burst_update(&tube->excitation, i);
    float y = 0.0;

    for (size_t i = 0; i < N_RESON; i++)
    {
        y += iirfilt_3_3_update(&tube->resonators[i], x);
    }

    return y / N_RESON;
}

chime_t *chime_init(void)
{
    chime_t *chime = &g_chime;
    chime->i = 0;
    trig_init(&chime->trig);
    chime->end = wind_init(&chime->trig.wind,
                           &(float[WIND_POINTS]){1.0 + get_rand_up(1.0),
                                                 2.0 + get_rand_up(2.0),
                                                 8.0 + get_rand_up(4.0),
                                                 2.0 + get_rand_up(3.0),
                                                 4.0 + get_rand_up(1.0)},
                           &(float[WIND_POINTS]){0.0, 2.0, 2.0, 0.0, 0.0});

    TUBES_INIT();

    return chime;
}

float chime_update(chime_t *chime)
{
    float y = 0.0;
    uint8_t s = trig_update(&chime->trig, chime->i);

    for (size_t i = 0; i < N_TUBES; i++)
    {
        y += tube_update(&chime->tubes[i], chime->i, s == (i + 1), trig_get_energy(&chime->trig));
    }
    chime->i++;

    return (y / N_TUBES) * 0.9;
}

bool chime_is_finished(chime_t *chime)
{
    return chime->i >= chime->end;
}
